primap2/pm2io/_data_reading.py

```diff
@@ -1414,33 +1414,65 @@ def preferred_unit(entity: str, units: list[str], gwp_to_use: Union[None, str])
 
     Examples
     --------
-    >>> preferred_unit("CO2", ["kt CO2 / yr", "mg CO2 / s"], None)
+    >>> preferred_unit("CO2", {"kt CO2 / yr": None, "mg CO2 / s": None})
     'Gg CO2 / yr'
-    >>> preferred_unit("CH4", ["kt CO2 / yr"], "AR4GWP100")
+    >>> preferred_unit("CH4", {"kt CO2 / yr": "AR4GWP100"})
     'Gg CH4 / yr'
-    >>> preferred_unit("CH4", ["kt CO2 / yr", "Mg CO2 / yr"], None)
+    >>> preferred_unit("CH4", {"kt CO2 / yr": None, "Mg CO2 / yr": None})
+    'kt CO2 / yr'
+     >>> preferred_unit("CH4", {"kt CO2 / yr": None, "Mg CH4 / yr": None})
+    None
+    >>> preferred_unit(
+    ...     "KYOTOGHG", {"kt CO2 / yr": "AR4GWP100", "Gg CO2 / yr": "SARGWP100"}
+    ... )
```

Comment:
Ah, it is rather:
> In a dataset, CH4 emissions are specified in GWP and in native units

right? Maybe I just need another example, like
```
    >>> preferred_unit(
    ...     "CH4", {"kt CO2 / yr": "AR4GWP100", "Gg CH4 / yr": None}
    ... )
```
to understand it better.

Reply:
That's actually a bug in the test. I wanted to write the test with a gas not a basket. With a basket it should fail as it's not convertible between gwps without additional assumptions. There's actually more of the doctests that pass but the result is a bit problematic (e.g. the third test where CH4 is given in CO2 eq without a GWP. This should throw an error instead of converting all to the first unit. It's technically correct that Mg CO2 / yr can be converted to kt CO2 / yr, but the unit is just nonsense for CH4 and maybe we should not read data in nonsensical units)

Reply:
Re-reading the function, I think the API is overcomplicated, isn't it? The only information that is needed to find out which unit should be preferred is if GWP metric conversions should be allowed or not. If that's the case, a better API could be:

```python
preferred_unit(entity: str, units: list[str], allow_gwp_conversions: bool) -> str:
```

That would be much easier to understand, at least for me.

Or what am I missing?

Reply:
And `is_compatible_with` always knows which conversions should be allowed, we pass this information in the conversion_contexts, no?

Reply:
Or do you actually mean metric conversions? Converting from "KYOTOGHG (SARGWP100)" to "KYOTOGHG (AR4GWP100)" is of course not possible without additional information, even though both have the same unit. Are metric conversions part of the question here? Because if so, we need a completely different answer. That's why the GWP metric specification is part of the `entity` in primap2. If you drop that information, things become underspecified.

Reply:
The functions is supposed to check if conversion to a native unit (no GWP) is possible and then prefers 'Gg / <entity> / yr'. Conversions between metrics / contexts are only done if possible without further knowledge / assumptions (e.g. CH4 in CO2eq,  SARGWP100 is converted to native units, but KYOTOGHG in CO2eq, SARGWP100 is not converted as there is no native unit for KYOTOGHG). That's all no problem. 
The problematic case is when we have a basket like KYOTOGHG given with different contexts. This is also not convertible without further knowledge and thus 'None' should be returned. The current code returns the first unit though, as 'is_compatible_with' knows about the contexts, but not about the entity. And the contexts allow conversion between each other for entities known to 'openscm_units', but not for baskets. But as 'is_compatible_with' doesn't know the entity and doesn't even know about the existence of entities where conversion between GWP contexts is not possible it returns 'True'. I've already fixed the code and am just fighting with the tests that complain that the functions returns nothing instead of 'None' (while it returns 'None' when I run it in the debugger)

Reply:
Yeah, that is actually what `harmonize_units` wants to do, right? Harmonize units and entities. Maybe it will help if we formulate the task for `harmonize_units` along those lines.

```python
def harmonize_emissions_units_and_entities(
    data: pd.DataFrame,
    *,
    unit_col: Union[None, str] = None,
    attrs: Union[None, dict] = None,
    dimensions: Iterable[str],
) -> None:
    """Harmonize the units and entities of the input GHG emissions data.

    For each gas, convert all time series to the same entity and unit. Prefers
    first to convert to native unit, i.e. mass of emitted substance per time.
    Sometimes, that is not possible, e.g. because we are dealing with a gas basket
    specified as a global warming potential, i.e. mass of CO2 which, if emitted,
    would lead to the same warming as the emitted substance per time.
    Then, simply the first unit used in the input data is preferred. Units
    must already be in PRIMAP2 style.

    Parameters
    ----------
    data: pd.DataFrame
        data for which the units should be harmonized
    unit_col: str, optional
        column name for unit column. Default: "unit"
    attrs: dict, optional
        attrs defining the aliasing of columns. If attrs contains "entity_terminology",
        "entity (<entity_terminology>)" will be used as the entity column, otherwise
        simply "entity" will be used as the entity column.
    dimensions: list of str
        the dimensions, i.e. the metadata columns.

    Returns
    -------
    None
        The data is altered in place.
    """
```

Note that if the input data contains a gas basket specified using two different GWP metrics, conversion is not possible with the available information, and we have to either not covert or raise an error.

As an aside, we probably will need this functionality in core primap2 over in xarray land. Therefore, we should try to build building blocks which can be used there as well to not have too much double work.

Reply:
Hadn't seen your last comment when I wrote my last comment.

Yeah, cool, then we arrived at a common understanding what should happen. The `None` vs. nothing is a special thing for doctests. doctests literally look at what happens on the python CLI and at the python CLI, `None` is shown as nothing. Technically, if you don't return anything from a function, `None` is returned, so actually returning `None` is the same as returning nothing:

```
~> python3
Python 3.11.2 (main, Mar 13 2023, 12:18:29) [GCC 12.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> def returns_int():
...     return 5
... 
>>> returns_int()
5
>>> def returns_none():
...     return None
... 
>>> def does_not_return():
...     pass
... 
>>> returns_none() is does_not_return()
True
>>> 
```

So, if you want to assert that `None` is returned in a doctest, you have to put nothing there.

Reply:
Or, do be more explicit, you can also in the doctests write
```
>>> function_under_test() is None
True
```

Reply:
It does work with an empty line as expected result. However, I have two tests for 'None' in the doctest for the function and the other one works with 'None'. They both get their return value from the same line in the code.